core/storage/trie_pruner/impl/trie_pruner_impl.hpp

```diff
@@ -0,0 +1,229 @@
+        std::shared_ptr<const storage::trie::Codec> codec,
+        std::shared_ptr<storage::SpacedStorage> storage,
+        std::shared_ptr<const crypto::Hasher> hasher);
+
+    virtual outcome::result<void> addNewState(
+        trie::PolkadotTrie const &new_trie,
+        trie::StateVersion version) override;
+
+    virtual outcome::result<void> addNewChildState(
+        storage::trie::RootHash const &parent_root,
+        trie::PolkadotTrie const &new_trie,
+        trie::StateVersion version) override;
+
+    virtual outcome::result<void> markAsChild(Parent parent,
+                                              Child child) override;
+
+    virtual outcome::result<void> pruneFinalized(
+        primitives::BlockHeader const &state,
+        primitives::BlockInfo const &next_block) override;
+
+    virtual outcome::result<void> pruneDiscarded(
+        primitives::BlockHeader const &state) override;
+
+    primitives::BlockNumber getBaseBlock() const override {
+      return base_block_.number;
+    }
+
+    size_t getTrackedNodesNum() const {
+      return ref_count_.size();
+    }
+
+    std::set<common::Buffer> generateTrackedNodeSet() const {
+      std::set<common::Buffer> set;
+      for (auto &[node, count] : ref_count_) {
+        set.insert(node);
+      }
+      return set;
+    }
+
+    size_t getRefCountOf(common::Buffer const &node) const {
+      auto it = ref_count_.find(node);
+      return it == ref_count_.end() ? 0 : it->second;
+    }
+
+    template <typename F>
+    void forRefCounts(F const &f) {
+      for (auto &[node, count] : ref_count_) {
+        f(node, count);
+      }
+    }
+
+    std::optional<uint32_t> getPruningDepth() const override {
+      return pruning_depth_;
+    }
+
+   private:
+    TriePrunerImpl(
+        std::optional<uint32_t> pruning_depth,
+        std::shared_ptr<storage::trie::TrieStorageBackend> trie_storage,
+        std::shared_ptr<const storage::trie::TrieSerializer> serializer,
+        std::shared_ptr<const storage::trie::Codec> codec,
+        std::shared_ptr<storage::SpacedStorage> storage,
+        std::shared_ptr<const crypto::Hasher> hasher)
+        : trie_storage_{trie_storage},
+          serializer_{serializer},
+          codec_{codec},
+          storage_{storage},
+          hasher_{hasher},
+          pruning_depth_{pruning_depth} {
+      BOOST_ASSERT(trie_storage_ != nullptr);
+      BOOST_ASSERT(serializer_ != nullptr);
+      BOOST_ASSERT(codec_ != nullptr);
+      BOOST_ASSERT(storage_ != nullptr);
+      BOOST_ASSERT(hasher_ != nullptr);
+    }
+
+    outcome::result<void> prune(primitives::BlockHeader const &state);
+
+    // policy of adding a new trie to the pruner ref counters
+    struct AddConfig {
+      enum AddType {
+        // register only the nodes which are already loaded in memory with the
+        // pruner (these are the nodes that were modified between this trie and
+        // its parent)
+        AddLoadedOnly,
```

Comment:
Not needed, because traversal will detect that hash was already indexed before reading it's node from db.
```suggestion
```